compiler/rustc_borrowck/src/type_check/mod.rs

```diff
@@ -622,6 +624,13 @@ impl<'a, 'b, 'tcx> TypeVerifier<'a, 'b, 'tcx> {
         swap_constraints(self);
 
         let locations = location.to_locations();
+
+        // Use location of promoted const in collected constraints
+        for type_test in type_tests.iter() {
```

Comment:
Is it possible to do this by iterating over `self.cx.borrowck_context.constraints.type_tests` directly and modifying in place? Or are the type tests shared with something else and the in place modification would break something else, or something like that?

Reply:
> Or are the type tests shared with something else and the in place modification would break something else, or something like that?

Yes the type tests are shared with the constraints of the borrowck context of the body that contains the promoted constant, so we need some way to only collect the constraints of the promoted constant body. [`swap_constraints`](https://github.com/rust-lang/rust/blob/72f4923979979abb5d6b975353e9b3053d257e60/compiler/rustc_borrowck/src/type_check/mod.rs#L595) is currently used for this. I'm not sure what the problem is with spawning an entirely new borrow checking context when type checking promoteds. Presumably this would cause slight performance problems and we only really care about the constraints being different? Not sure.